src/tree.rs

```diff
@@ -1,22 +1,138 @@
 use lazy::Lazy;
+use std::fmt::{Debug, Error, Formatter};
+use std::rc::Rc;
 
+#[derive(Clone)]
 pub struct Tree<'a, A> {
     thunk: Lazy<'a, A>,
-    #[allow(dead_code)]
-    children: Vec<Tree<'a, A>>,
+    pub children: Vec<Tree<'a, A>>,
 }
 
-impl<'a, A: 'a + Clone> Tree<'a, A> {
+impl<'a, A> Tree<'a, A>
+where
+    A: 'a + Clone,
+{
+    fn new(value: A, children: Vec<Tree<'a, A>>) -> Self {
+        let thunk = Lazy::new(value);
+        Tree { thunk, children }
+    }
+
     pub fn singleton(value: A) -> Tree<'a, A> {
         Tree {
-            thunk: Lazy::new(move || value.clone()),
+            thunk: Lazy::new(value),
             children: vec![],
         }
     }
 
-    pub fn value(&mut self) -> &A {
+    pub fn value(&self) -> Option<A> {
         self.thunk.value()
     }
+
+    pub fn expand<F>(f: Rc<F>) -> impl Fn(Tree<'a, A>) -> Tree<'a, A>
+    where
+        F: Fn(A) -> Vec<A>,
+    {
+        move |t: Tree<'a, A>| {
+            let mut children: Vec<Tree<'a, A>> = t
+                .children
+                .iter()
+                .map(|t| Self::expand(f.clone())(t.clone()))
+                .collect();
+            let mut zs = unfold_forest(Rc::new(move |x| x), f.clone(), t.value().unwrap());
+            children.append(&mut zs);
+            Tree::new(t.value().unwrap(), children)
+        }
+    }
+}
+
+// TODO: Needs to accept an Rc'd F.
+pub fn bind<'a, A, B, F>(t: Tree<'a, A>) -> impl Fn(Rc<F>) -> Tree<'a, B>
+where
+    A: Clone + 'a,
+    B: Clone + 'a,
+    F: Fn(A) -> Tree<'a, B> + 'a,
+{
+    let x = t.value();
+    let xs0 = t.children;
+    move |k: Rc<F>| {
+        let mut t1 = k(x.clone().unwrap());
+        let mut xs: Vec<Tree<'a, B>> = xs0.iter().map(|m| bind(m.clone())(k.clone())).collect();
+        xs.append(&mut t1.children);
+        Tree {
+            thunk: t1.thunk,
+            children: xs,
+        }
+    }
+}
+
+impl<'a, A> Debug for Tree<'a, A>
+where
+    A: 'a + Clone + Debug,
+{
+    fn fmt(&self, f: &mut Formatter) -> Result<(), Error> {
+        // TODO: Actually print the children.
+        let has_children_str = if self.children.is_empty() {
+            format!("<>")
+        } else {
+            format!("<children>")
```

Comment:
being lazy for the moment, although theoretically this is a simple implementation; you just go through each child and call fmt on them.

Reply:
this can't be derived automatically?

Reply:
Fixed in 4a8e554